### ISSUE TITLE
Catch Wrong/Reversed order when retrieving messages from archive...

### ISSUE THREAD

**Issue by mightymop** [jsxc/jsxc#1002]
### Expected behavior
When using "Load old messages" the ordering of messages should be the same as it was when the messages were written before...

Message order should be (top to bottom) A, B, C, D, E, F
### Actual behavior
When i try to retrieve message from archive via xep313 ("Load old messages" Button) the oldest message is on bottom and the newest on top.
Message order is (top to bottom) F, E, D, C, B, A

### Steps to reproduce the behavior
1. login
2. open chatbox
3. click on "load old messages" button

### Environment
- **JSXC version:** actual master, but same problem on your demo app
- **Host system and version:** ...
- **Browser vendor and version:** Firefox and Chrome
- **Any browser plugins enabled?**  no
- **XMPP server vendor and version:** Openfire 4.6 & Message Archivemenagement Plugin 2.2.0
- **Is your XMPP server working with other clients as expected?** Yes Converse and Gajim show messages in right order

#### Javascript

```
no error messages or messages related to mam other than the following in xmpp section
```

#### JSXC

```
no error messages or messages related to mam other than the following in xmpp section
```

#### XMPP

This comes from the server log
```
<iq id="b74c8082-54a4-4770-b5ef-fa5658665f10:sendIQ" to="mops@mydomain.local" type="set" from="mops@mydomain.local/jsxc-b69642ac"><query xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019"><x xmlns="jabber:x:data" type="submit"><field type="hidden" var="FORM_TYPE"><value>urn:xmpp:mam:2</value></field><field var="with"><value>admin@mydomain.local</value></field></x><set xmlns="http://jabber.org/protocol/rsm"><max>20</max><before/></set></query></iq>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3263"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:57.372Z"/><message xmlns="jabber:client" id="131f9520-ac13-4cff-b65e-baa8d28c7b90" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>F</body><origin-id xmlns="urn:xmpp:sid:0" id="131f9520-ac13-4cff-b65e-baa8d28c7b90"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3262"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:55.759Z"/><message xmlns="jabber:client" id="97b11150-d4d8-4ed3-b5c0-955954d3f91f" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>E</body><origin-id xmlns="urn:xmpp:sid:0" id="97b11150-d4d8-4ed3-b5c0-955954d3f91f"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3261"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:54.186Z"/><message xmlns="jabber:client" id="b133b07c-9c8e-4320-99a2-fadff47dd281" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>D</body><origin-id xmlns="urn:xmpp:sid:0" id="b133b07c-9c8e-4320-99a2-fadff47dd281"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3260"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:52.670Z"/><message xmlns="jabber:client" id="49b15b84-b6de-4285-b100-01bff702f0ab" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>C</body><origin-id xmlns="urn:xmpp:sid:0" id="49b15b84-b6de-4285-b100-01bff702f0ab"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3259"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:51.217Z"/><message xmlns="jabber:client" id="2e9ba264-4f01-4de8-830f-f676b33c7746" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>B</body><origin-id xmlns="urn:xmpp:sid:0" id="2e9ba264-4f01-4de8-830f-f676b33c7746"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3258"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:50.371Z"/><message xmlns="jabber:client" id="7ac40b3d-d209-461f-ac02-341414e9d52c" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>A</body><origin-id xmlns="urn:xmpp:sid:0" id="7ac40b3d-d209-461f-ac02-341414e9d52c"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<iq type="result" id="b74c8082-54a4-4770-b5ef-fa5658665f10:sendIQ" from="mops@mydomain.local" to="mops@mydomain.local/jsxc-b69642ac"><fin xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019"><set xmlns="http://jabber.org/protocol/rsm"><first>3263</first><last>3258</last><count>6</count></set></fin></iq>
```

i noticed the id (3263) in result iq the in "&lt;first&gt;" is higher than the id (3258) in "&lt;last&gt;"
&lt;set xmlns="http://jabber.org/protocol/rsm" &gt; &lt;first&gt;3263&lt;/first&gt;&lt;last&gt;3258&lt;/last&gt;&lt;count&gt;6&lt;/count&gt;&lt;/set&gt;...

May be its related to the servers implementation and depends on database query result ordering (asc or desc)... but this could be catched if we compare the &lt;first&gt; tag with &lt;last&gt; tag and do reverse the messages if &lt;first&gt; is higher than &lt;last&gt;

Further more because of the wrong order every further click on "load old messages" will result in a corrupted message order in the chatbox then...

maybe this issue https://github.com/jsxc/jsxc/issues/999 is related to this here too...

**Comment by sualko** [comment]
If I remember correctly the order is defined in the mam xep, but I have to look it up.

The result id also doesn't have to be numeric. Prosody e.g. is using a random string.

**Comment by mightymop** [comment]
> 
> 
> If I remember correctly the order is defined in the mam xep, but I have to look it up.
> 
> The result id also doesn't have to be numeric. Prosody e.g. is using a random string.

hm ok then i will compare the last tag with last index item in messageCache... thanks...

**Comment by mightymop** [comment]
The bug in OF MAM plugin is now fixed in latest Release 2.2.1. So now this is a minor improval if we would add this catch
